File: 3.5/2.0/LinFu.DesignByContract2.Examples/VBContractAssertions/Assertions/Postconditions/EnsureBalanceReflectsWithdrawalAmountAttribute.vb
Imports LibraryInterfaces
Imports LinFu.DesignByContract2.Core

<AttributeUsage(AttributeTargets.Method Or AttributeTargets.ReturnValue)> _
Public Class EnsureBalanceReflectsWithdrawalAmountAttribute
    Inherits Attribute
    Implements IPostcondition
    Private oldBalance As Integer
    Private expectedBalance As Integer


    Public Function AppliesTo(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo) As Boolean Implements LinFu.DesignByContract2.Core.IContractCheck.AppliesTo
        If target Is Nothing Then Return False
        If Not (TypeOf target Is IBankAccount) Then Return False
        Return True
    End Function

    Public Sub [Catch](ByVal ex As System.Exception) Implements LinFu.DesignByContract2.Core.IContractCheck.Catch

    End Sub

    Public Sub BeforeMethodCall(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo) Implements LinFu.DesignByContract2.Core.IPostcondition.BeforeMethodCall
        If target Is Nothing Then Return
        If Not (TypeOf target Is IBankAccount) Then Return

        Dim account As IBankAccount = CType(target, IBankAccount)
        oldBalance = account.Balance
    End Sub

    Public Function Check(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal returnValue As Object) As Boolean Implements LinFu.DesignByContract2.Core.IPostcondition.Check
        If target Is Nothing Then Return True
        If Not (TypeOf target Is IBankAccount) Then Return True

        Dim account As IBankAccount = CType(target, IBankAccount)
        Dim amount As Integer = CType(info.Arguments(0), Integer)
        expectedBalance = oldBalance - amount
        Return account.Balance = expectedBalance
    End Function

    Public Sub ShowError(ByVal output As System.IO.TextWriter, ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal returnValue As Object) Implements LinFu.DesignByContract2.Core.IPostcondition.ShowError
        output.WriteLine("Withdrawal Failed! The expected balance should have been '{0}'", expectedBalance)
    End Sub
End Class

File: 3.5/2.0/LinFu.DesignByContract2.Examples/ContractAssertions/Assertions/Postconditions/EnsureBalanceReflectsWithdrawalAmountAttribute.cs
using System;
using System.Collections.Generic;
using System.Text;
using LibraryInterfaces;
using LinFu.DesignByContract2.Core;
using LinFu.DynamicProxy;

namespace CSContractAssertions
{
    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
    public class EnsureBalanceReflectsWithdrawalAmountAttribute : Attribute, IPostcondition
    {
        private int _oldBalance;
        private int _expectedBalance;
        #region IPostcondition Members

        public void BeforeMethodCall(object target, InvocationInfo info)
        {
            if (target == null || !(target is IBankAccount))
                return;
            IBankAccount account = (IBankAccount)target;
            _oldBalance = account.Balance;
        }

        public bool Check(object target, InvocationInfo info, object returnValue)
        {
            if (target == null || !(target is IBankAccount))
                return true;

            IBankAccount account = target as IBankAccount;
            int amount = (int)info.Arguments[0];
            _expectedBalance = _oldBalance - amount;

            return account.Balance == _expectedBalance;
        }

        public void ShowError(System.IO.TextWriter output, object target, InvocationInfo info, object returnValue)
        {
            output.WriteLine("Withdrawal Failed! The expected balance should have been '{0}'", _expectedBalance);
        }

        #endregion

        #region IContractCheck Members

        public bool AppliesTo(object target, InvocationInfo info)
        {
            return target is IBankAccount;
        }

        public void Catch(Exception ex)
        {
            // Do nothing
        }

        #endregion
    }
}
